src/liballoc/sync.rs

```diff
@@ -1034,14 +1031,28 @@ impl<T> Weak<T> {
     /// ```
     #[stable(feature = "downgraded_weak", since = "1.10.0")]
     pub fn new() -> Weak<T> {
-        unsafe {
-            Weak {
-                ptr: NonNull::new_unchecked(WEAK_EMPTY as *mut _),
-            }
+        Weak {
+            ptr: <ArcInner<T>>::dangling_ptr(),
         }
     }
 }
 
+// Specialization because `NonNull::<T>::dangling()` only exits for `T: Sized`
```

Comment:
Why not allow `NonNull::<T>::dangling()` for `T: ?Sized` instead?

Reply:
It wouldn’t compile because `usize` can only be cast to thin pointers because we wouldn’t know what to use for the vtable pointer (for trait objects, for example).

Reply:
Oh, wait, I see now... this panics when actually called with something unsized? Is that really the desired behavior, for `std::rc::Weak::<str>::new()` to panic?

Reply:
Sorry, I definitely should have commented this more clearly.

Yes, the idea was that `Weak::new` only exists for `T: Sized`. And yes, this is broken because I didn’t think of unsizing coecion. Good catch, thanks.